### PR TITLE
[IRGen] Mix -profile-generate into module hash

### PR DESCRIPTION
**Pull request by beccadax**
When rebuilding a .o file, we hash the IR generated by IRGen and compare it to a hash embedded in the .o file. If they match, we simply skip asking LLVM to optimize and generate code. Certain flags that control our LLVM configuration are mixed into the hash, but the -profile-generate flag was not one of them. This usually wouldn’t matter because profiling would insert additional code into the IR, but in edge cases like empty files or files containing only protocol declarations, it could cause linker errors when profiling was turned off.

This change adds the `GenerateProfile` flag into the IR hash, ensuring that we always recompile, even if the IR was identical.

Fixes rdar://problem/54126622.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by gottesmm** [comment]
This is for @eeckstein

**Comment by vedantk** [comment]
Thanks for working on this. To be on the safe side, should other types of instrumentation be included here as well (say, {a,t}san)?

**Review by eeckstein — APPROVED** [review]
lgtm!

**Comment by gottesmm** [comment]
Question. Would it be possible to change this into some sort of exhaustive switch for Options? That way if someone adds a new option like this, they get a warning telling them that they may need to update this code.

**Comment by gottesmm** [comment]
In other words, I am worried that the code as written will cause this same bug to occur again. If we had some sort of switch associated with it, at least we could get a compile time warning in such a case. This will cause the writer to at least think about... should this be used in the IR hash?

**Comment by jrose-apple** [comment]
It's only required for options that don't affect the IR as generated, but arguably being forced to say that explicitly would be nice. You could do this with a .def file but that might feel heavyweight.

**Comment by gottesmm** [comment]
@jrose-apple exactly. My fear is that someone will just add the option and not think about it. But if one at least has to explicitly do something, that can not happen.

**Comment by brentdax** [comment]
@vedantk Good idea to check. `-sanitize=address` and `-sanitize=thread` emit different IR for an empty source file. `-sanitize=undefined` emits the same IR, but also the same assembly on x86_64 at least. `-sanitize=fuzzer` emits the same IR but different assembly, so that one should be mixed in too.

**Comment by brentdax** [comment]
@gottesmm @jrose-apple I don't think the problem here is that someone forgot to update the function when they added an option; I think it's that someone didn't think about this edge case of an empty file, or didn't know about LLVM's behavior in this situation. `-profile-generate` doesn't usually need to be mixed into the flags because it will normally cause IR changes; only in files with no executable code does it behave incorrectly.

**Comment by jrose-apple** [comment]
We could do that in +Asserts builds. We're only losing incrementality by doing so.

**Comment by brentdax** [comment]
This is probably the most complicated and expensive "assertion" I've ever written, but it's in and it works.

**Comment by brentdax** [comment]
@swift-ci please test

File: lib/IRGen/IRGen.cpp [hunk 1]
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
...
+                                          OriginalOutputFilename,
+                                          /*tolerances=*/0, 0, &message)) {
+      if (DiagMutex)
+        DiagMutex->lock();
+      llvm::errs() << "Object file comparison failed: " << message
+                   << "\n  Old file: " << OriginalOutputFilename
+                   << "\n  New file: " << OutputFilename << "\n";
+      if (DiagMutex)
+        DiagMutex->unlock();
+      llvm_unreachable(
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: Nitpick: This should use `llvm::report_fatal_error`.

File: lib/IRGen/IRGen.cpp [hunk 2]
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
...
+  if (!OriginalOutputFilename.empty()) {
+    // We're done changing the file; make sure it's saved before we compare.
+    RawOS->close();
+
+    std::string message;
+    if (0 != llvm::DiffFilesWithTolerance(OutputFilename,
+                                          OriginalOutputFilename,
+                                          /*tolerances=*/0, 0, &message)) {
+      if (DiagMutex)
+        DiagMutex->lock();
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: I don't think you need to take the mutex here (or below); you're just going to crash anyway, and the temporary file doesn't affect other threads.

File: lib/IRGen/IRGen.cpp [hunk 3]
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
...
+
+#ifndef NDEBUG
+  if (!OriginalOutputFilename.empty()) {
+    // We're done changing the file; make sure it's saved before we compare.
+    RawOS->close();
+
+    std::string message;
+    if (0 != llvm::DiffFilesWithTolerance(OutputFilename,
+                                          OriginalOutputFilename,
+                                          /*tolerances=*/0, 0, &message)) {
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: This seems like overkill. Check out `swift::moveFileIfDifferent`.

File: lib/IRGen/IRGen.cpp [hunk 4]
```diff
@@ -561,6 +591,45 @@ bool swift::performLLVM(IRGenOptions &Opts, DiagnosticEngine *Diags,
...
+      // that changed LLVM's output between this compile and the previous one.
+    } else {
+      if (auto EC = llvm::sys::fs::remove(OutputFilename)) {
+        if (DiagMutex)
+          DiagMutex->lock();
+        llvm::errs() << "Cannot remove temp file '" << OutputFilename << "': "
+                     << EC.message() << "\n";
+        if (DiagMutex)
+          DiagMutex->unlock();
+        llvm_unreachable("Unexpected error deleting temp file");
```
Comment (jrose-apple) [inline lib/IRGen/IRGen.cpp]: Here too…or we could just ignore it.

**Review by jrose-apple — COMMENTED** [review]
Nice work. cc @eeckstein

File: lib/IRGen/IRGen.cpp (hunk 2 above)
Comment (brentdax) [reply lib/IRGen/IRGen.cpp]: The mutex is to keep several threads from writing messages at the same time (or at least that's my reading of why it's done elsewhere in the file), but I think I'll simplify the output a bit anyway.

File: include/swift/AST/IRGenOptions.h [hunk 5]
```diff
@@ -259,6 +259,8 @@ class IRGenOptions {
     unsigned Hash = (unsigned)OptMode;
     Hash = (Hash << 1) | DisableLLVMOptzns;
     Hash = (Hash << 1) | DisableSwiftSpecificLLVMOptzns;
+    Hash = (Hash << 1) | GenerateProfile;
+    Hash = (Hash << 1) | Sanitizers.contains(SanitizerKind::Fuzzer);
```
Comment (jrose-apple) [inline include/swift/AST/IRGenOptions.h]: Honestly we might as well mix in the whole thing. It's not just about empty files; it's also about *non-*empty files that only change due to LLVM passes, and I'm not sure I trust that all the other sanitizers don't need this.

File: test/IRGen/Inputs/empty.swift [hunk 6]
```diff
@@ -0,0 +1 @@
+// This space intentionally left blank.
```
Comment (jrose-apple) [inline test/IRGen/Inputs/empty.swift]: Nitpick: There's already a test/Inputs/empty.swift you can use.

**Comment by jrose-apple** [comment]
Thanks for the refactor!

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
@jrose-apple I mixed in all of the sanitizers, and also redesigned the hash slightly to avoid any issues when we add more sanitizers/optimization modes/values to these fields in general.

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Pull request merged**